video/external/src/decoder/webcodecs.rs

```diff
@@ -0,0 +1,277 @@
+    length_size: u8,
+
+    /// The WebCodecs decoder object.
+    decoder: WebVideoDecoder,
+
+    /// The decoder output callback writes this, and the decode_frame method reads it.
+    ///
+    /// This in itself results in one frame of delay (because we can't block decode_frame
+    /// until the callback is invoked), but it shouldn't matter in practice.
+    last_frame: Rc<RefCell<Option<DecodedFrame>>>,
+
+    // Simply keeping these objects alive, as they are used by the JS side.
+    // See: https://rustwasm.github.io/wasm-bindgen/examples/closures.html
+    #[allow(dead_code)]
+    output_callback: Closure<dyn Fn(VideoFrame)>,
+    #[allow(dead_code)]
+    error_callback: Closure<dyn Fn(DomException)>,
+}
+
+impl H264Decoder {
+    /// `extradata` should hold "AVCC (MP4) format" decoder configuration, including PPS and SPS.
+    /// Make sure it has any start code emulation prevention "three bytes" removed.
+    ///
+    /// The log_subscriber is needed so that we have proper logging from within the callbacks.
+    pub fn new(log_subscriber: Arc<Layered<WASMLayer, Registry>>) -> Result<Self, Error> {
+        let last_frame = Rc::new(RefCell::new(None));
+        let lf = last_frame.clone();
+
+        let output = move |output: &VideoFrame| {
+            let _subscriber = tracing::subscriber::set_default(log_subscriber.clone());
+            let visible_rect = output.visible_rect().unwrap();
+
+            match output.format().unwrap() {
+                VideoPixelFormat::I420 => {
+                    let mut data: Vec<u8> =
+                        vec![
+                            0;
+                            visible_rect.width() as usize * visible_rect.height() as usize * 3 / 2
+                        ];
+                    let _ = output.copy_to_with_u8_slice(&mut data);
+                    last_frame.replace(Some(DecodedFrame::new(
+                        visible_rect.width() as u32,
+                        visible_rect.height() as u32,
+                        BitmapFormat::Yuv420p,
+                        data,
+                    )));
+                }
+                VideoPixelFormat::Bgrx => {
+                    let mut data: Vec<u8> =
+                        vec![0; visible_rect.width() as usize * visible_rect.height() as usize * 4];
+                    let _ = output.copy_to_with_u8_slice(&mut data);
+                    for pixel in data.chunks_mut(4) {
+                        pixel.swap(0, 2);
+                        pixel[3] = 0xff;
+                    }
+                    last_frame.replace(Some(DecodedFrame::new(
+                        visible_rect.width() as u32,
+                        visible_rect.height() as u32,
+                        BitmapFormat::Rgba,
+                        data,
+                    )));
+                }
+                _ => {
+                    panic!("unsupported pixel format: {:?}", output.format().unwrap());
```

Comment:
Admittedly I'm not great with Rust, but does it make sense to panic if this is any other format?

https://developer.mozilla.org/en-US/docs/Web/API/VideoFrame/format and https://rustwasm.github.io/wasm-bindgen/api/web_sys/enum.VideoPixelFormat.html lists a lot more formats than just I420 and BGRX. Is it possible to embed a video in any of those other formats with Flash, and if so, wouldn't the whole thing panic when trying to play them.

Reply:
Yeah, good point, this is not nice.